src/main/java/com/snowflake/kafka/connector/internal/ingestsdk/StreamingClientManager.java

```diff
@@ -0,0 +1,147 @@
+/*
+ * Copyright (c) 2023 Snowflake Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package com.snowflake.kafka.connector.internal.ingestsdk;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.snowflake.kafka.connector.Utils;
+import com.snowflake.kafka.connector.internal.LoggerHandler;
+import com.snowflake.kafka.connector.internal.SnowflakeErrors;
+import com.snowflake.kafka.connector.internal.streaming.StreamingUtils;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Properties;
+
+/**
+ * Provides access to the streaming ingest clients. This should be the only place to manage clients.
+ */
+public class StreamingClientManager {
+  private LoggerHandler LOGGER;
+
+  private Map<Integer, KcStreamingIngestClient> taskToClientMap;
+  private int maxTasks;
+  private final int minTasks = 0;
+
+  // TESTING ONLY - inject the client map
+  @VisibleForTesting
+  public StreamingClientManager(Map<Integer, KcStreamingIngestClient> taskToClientMap) {
+    this();
+    this.taskToClientMap = taskToClientMap;
+  }
+
+  /** Creates a new client manager */
+  protected StreamingClientManager() {
+    LOGGER = new LoggerHandler(this.getClass().getName());
+    this.taskToClientMap = new HashMap<>();
+    this.maxTasks = 0;
+  }
+
+  /**
+   * Creates as many clients as needed with the connector config and kc instance id. This assumes
+   * that all taskIds are consecutive ranging from 0 -> maxTasks.
+   *
+   * @param connectorConfig the config for the clients, cannot be null
+   * @param kcInstanceId the kafka connector id requesting the clients, cannot be null
+   * @param maxTasks the max number of tasks assigned to this connector, must be > 0
+   * @param numTasksPerClient the max number of tasks to be assigned to each client, must be > 0
+   */
+  public void createAllStreamingClients(
+      Map<String, String> connectorConfig,
+      String kcInstanceId,
+      int maxTasks,
+      int numTasksPerClient) {
```

Comment:
Who controls these args? Is it possible to consolidate them so the below can be simplified?

Reply:
maxTasks comes from the config. numTasksPerClient is a constant set in SnowflakeSinkConnector.java for perf testing. We do need some logic to load balance tasks across clients, this was the simplest way i could come up with

`  private static final int NUM_TASK_TO_CLIENT = 1;`

Reply:
Ack, makes sense.